### PR TITLE
[ResourceBundle] the target resolver have been reworked (+ mongo support)

### PR DESCRIPTION
**Pull request by aRn0D**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by amenophis** [comment]
:+1:

File: src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php [hunk 1]
```diff
@@ -42,19 +40,42 @@ public function resolve(ContainerBuilder $container, array $interfaces)
...
+     * @return string
+     */
+    abstract protected function getTagName();
+
+    /**
+     * Return the name of the method which is called on the listener
+     *
+     * @return string
+     */
+    abstract protected function getMethodName();
```
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php]: Why not const ?

File: src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php [hunk 2]
```diff
@@ -42,19 +40,42 @@ public function resolve(ContainerBuilder $container, array $interfaces)
...
+
+    /**
+     * Return the doctrine resolve target listener
+     *
+     * @return string
+     */
+    abstract protected function getTargetResolverService();
```
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php]: Why not const ?

File: src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php [hunk 3]
```diff
@@ -42,19 +40,42 @@ public function resolve(ContainerBuilder $container, array $interfaces)
...
+            $resolveTargetEntityListener->addTag($tagName, array('event' => 'loadClassMetadata'));
         }
     }
 
     /**
+     * Return the tag name
+     *
+     * @return string
+     */
+    abstract protected function getTagName();
```
Comment (amenophis) [inline src/Sylius/Bundle/ResourceBundle/Doctrine/Common/AbstractTargetResolver.php]: Why not const ?

**Comment by aRn0D** [comment]
@pjedrzejewski What do you think?

**Comment by pjedrzejewski** [comment]
Looks very cool, just an idea - We introduced these resolvers because prepend was not available yet. Now we can use the container prepend extension interface to simply preconfigure the interface -> class mapping in config. What do you think?

**Comment by aRn0D** [comment]
@pjedrzejewski  yes you're right! I will rework it !

**Comment by aRn0D** [comment]
@pjedrzejewski did you speak about the semantic config? 

For now, we have two kind a configurations (basic and advanced). Their structure are different, it is not consistent (and I don't like that). I want to change that but it requires a big refactoring (and big BC breaks). We need to share stuff between `AbstractResourceExtension` and `SyliusResourceExtension`. `AbstractResourceExtension` become fat, I want to extract some stuff from it, create extra classes (and them ;)). 

What do you think about this changes ?

**Comment by aRn0D** [comment]
@pjedrzejewski I am not sure it is possibe "An Extension can prepend the configuration of any Bundle before the load() method is called by implementing PrependExtensionInterface:"

**Comment by Shine-neko** [comment]
It going ?

**Comment by pjedrzejewski** [comment]
@aRn0D Any ideas? What is the status of this?

**Comment by Shine-neko** [comment]
@aRn0D ?

**Pull request closed without merging**